prog/dftb+/prg_dftb/initprogram.F90

```diff
@@ -1943,36 +1947,55 @@ subroutine initProgramVariables(input, env)
-                  & sum(qInput(1:orb%nOrbAtom(ii),ii,1))
+              qInput(1:orb%nOrbAtom(ii),ii,2) = qInput(1:orb%nOrbAtom(ii),ii,1) *&
+                  & input%ctrl%initialSpins(1,ii) / sum(qInput(1:orb%nOrbAtom(ii),ii,1))
+            end do
+          else if (input%ctrl%randomInitialSpins) then
+            do ii = 1, nAtom
+              call getRandom(randomSpins,rTmp)
+              rTmp = 2.0_dp * (rTmp - 0.5_dp)
+              qInput(1:orb%nOrbAtom(ii),ii,2) = rTmp * qInput(1:orb%nOrbAtom(ii),ii,1)
             end do
           else
             do ii = 1, nAtom
-              qInput(1:orb%nOrbAtom(ii),ii,2) = &
-                  & qInput(1:orb%nOrbAtom(ii),ii,1) * &
+              qInput(1:orb%nOrbAtom(ii),ii,2) = qInput(1:orb%nOrbAtom(ii),ii,1) *&
                   & (nEl(1)-nEl(2))/sum(qInput(:,:,1))
             end do
           end if
         case (4)
+          ! non-collinear spins and/or spin orbit
           if (tSpin) then
-            if (.not. allocated(input%ctrl%initialSpins)) then
+            ! spins
+            if (.not. (allocated(input%ctrl%initialSpins) .or. input%ctrl%randomInitialSpins)) then
               call error("Missing initial spins!")
             end if
-            if (any(shape(input%ctrl%initialSpins)/=(/3,nAtom/))) then
-              call error("Incorrect shape initialSpins array!")
-            end if
-            do ii = 1, nAtom
-              do jj = 1, 3
-                qInput(1:orb%nOrbAtom(ii),ii,jj+1) = &
-                    & qInput(1:orb%nOrbAtom(ii),ii,1) * &
-                    & input%ctrl%initialSpins(jj,ii) &
-                    & / sum(qInput(1:orb%nOrbAtom(ii),ii,1))
+            if (input%ctrl%randomInitialSpins) then
+              do ii = 1, nAtom
+                ! spin direction in as spherical polar angles
+                call getRandom(randomSpins, qInput(:orb%nOrbAtom(ii),ii,2:3))
+                qInput(:orb%nOrbAtom(ii),ii,2:3) = 2.0_dp* pi * qInput(:orb%nOrbAtom(ii),ii,2:3)
+                do jj = 1, orb%nOrbAtom(ii)
+                  ! fraction of charge
+                  call getRandom(randomSpins,rTmp)
+                  qInput(jj,ii,2:4) = rTmp * qInput(jj,ii,1) *&
+                      & [ sin(qInput(jj,ii,2))*cos(qInput(jj,ii,3)),&
+                      & sin(qInput(jj,ii,2))*sin(qInput(jj,ii,3)), cos(qInput(jj,ii,2)) ]
+                end do
               end do
-            end do
+            else
+              if (any(shape(input%ctrl%initialSpins)/=(/3,nAtom/))) then
+                call error("Incorrect shape initialSpins array!")
+              end if
+              do ii = 1, nAtom
+                do jj = 1, 3
+                  qInput(1:orb%nOrbAtom(ii),ii,jj+1) = qInput(1:orb%nOrbAtom(ii),ii,1) *&
+                      & input%ctrl%initialSpins(jj,ii) / sum(qInput(1:orb%nOrbAtom(ii),ii,1))
+                end do
+              end do
+            end if
           end if
         end select
+
```

Comment:
At some point, we will have to package initprogram also in subroutines, as it became a real mess. Maybe it would worth to package the charge/spin initialisation as a separate routine?